java/arcs/android/sdk/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,394 @@
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.CoroutineContext
+import kotlin.reflect.KClass
+
+/**
+ * Tool which can be used by [ArcHost]s to handle [Intent] based API calls, as well as
+ * send API calls.
+ *
+ * ## Example Usage:
+ *
+ * ```kotlin
+ * class MyService : Service() {
+ *     private val myHelper: ArcHostHelper by lazy {
+ *         ArcHostHelper(MyArcHost())
+ *     }
+ *
+ *     class MyArcHost : AbstractArcHost() {
+ *         override suspend fun onStartArc(plan: PlanPartition) {
+ *             // ...
+ *         }
+ *     }
+ *
+ *     override fun onStartCommand(intent: Intent?, flags: Int, startId: Int) {
+ *         val result = super.onStartCommand(intent, flags, startId)
+ *         myHelper.onStartCommand(intent)
+ *         return result
+ *     }
+ * }
+ * ```
+ */
+class ArcHostHelper(
+    private val service: Service,
+    private val arcHost: ArcHost,
+    private val coroutineContext: CoroutineContext = Dispatchers.IO
+) {
+
+    private val job = Job() + coroutineContext + CoroutineName("ArcHostHelper")
+
+    /**
+     * Determines whether or not the given [Intent] represents a call to an [ArcHost] and invokes
+     * the appropriate interface methods.
+     */
+    fun onStartCommand(intent: Intent?) = CoroutineScope(job).launch {
+        onStartCommandSuspendable(intent)
+    }
+
+    @VisibleForTesting
+    suspend fun onStartCommandSuspendable(intent: Intent?) {
+        // Ignore other actions
+        if (intent?.action?.startsWith(
+                ARC_HOST_INTENT
+            ) != true
+        ) return
+
+        // Ignore Intent when it doesn't target our Service
+        if (intent.component?.equals(ComponentName(service, service::class.java)) != true) return
+
+        when (intent.getIntExtra(
+            OPERATION, Operation.values().size
+        ).toOperation()) {
+            Operation.START_ARC -> runWithResult(
+                intent, ParcelablePlanPartition::class, arcHost::startArc
+            )
+            Operation.STOP_ARC -> runWithResult(
+                intent, ParcelablePlanPartition::class, arcHost::stopArc
+            )
+            Operation.REGISTER_PARTICLE -> runWithResult(
+                intent, ParcelableParticleIdentifier::class, arcHost::registerParticle
+            )
+            Operation.UNREGISTER_PARTICLE -> runWithResult(
+                intent, ParcelableParticleIdentifier::class, arcHost::unregisterParticle
+            )
+            Operation
+                .GET_REGISTERED_PARTICLES -> runWithResult(intent, arcHost::registeredParticles)
+            else -> {
+                throw IllegalArgumentException("Operation $intent not implemented.")
+            }
```

Comment:
It'd probably be best to leave this out, so a compilation error happens when a new `Operation` is created, but not implemented.